compiler/internal/typeparams/collect.go

```diff
@@ -131,21 +192,65 @@ func (c *visitor) Visit(n ast.Node) (w ast.Visitor) {
 	if t, ok := typ.(*types.Named); ok {
 		obj = t.Obj()
 	}
+
+	// If the object is defined in the same scope as the instance,
+	// then we apply the current nested type arguments.
+	var tNest []types.Type
+	if obj.Parent().Contains(ident.Pos()) {
+		tNest = c.tNest
+	}
+
+	c.addInstance(obj, tArgs, tNest)
+}
+
+func (c *visitor) visitNestedType(obj types.Object) {
+	if _, ok := obj.(*types.TypeName); !ok {
+		// Found a variable or function, not a type, so skip it.
+		return
+	}
+
+	typ := obj.Type()
+	if ptr, ok := typ.(*types.Pointer); ok {
+		typ = ptr.Elem()
+	}
+
+	t, ok := typ.(*types.Named)
+	if !ok || t.TypeParams().Len() > 0 {
+		// Found a generic type or an unnamed type (e.g. type parameter).
+		// Don't add generic types yet because they
+		// will be added when we find an instance of them.
+		return
+	}
+
+	c.addInstance(obj, nil, c.resolver.TypeArgs())
+}
+
+func (c *visitor) addInstance(obj types.Object, tArgList *types.TypeList, tNest []types.Type) {
+	tArgs := c.resolver.SubstituteAll(tArgList)
+	if isGeneric(tArgs...) {
+		// Skip any instances that still have type parameters in them after
+		// substitution. This occurs when a type is defined while nested
+		// in a generic context and is not fully instantiated yet.
+		// We need to wait until we find a full instantiation of the type.
```

Comment:
This also happens when the type argument itself is a nested type that hasn't been instantiated correctly. This is one of the fixes that I'm doing as a follow up PR. See "Nested Type Arguments" in summary for more information.